pymongo/database.py

```diff
@@ -541,36 +541,66 @@ def command(self, command, value=1, check=True,
+
+        if batch_size:
+            cmd["cursor"]["batchSize"] = batch_size
+
+        if filter:
+            cmd["filter"] = filter
 
         if sock_info.max_wire_version > 2:
             coll = self["$cmd"]
             with self.__client._tmp_session(session, close=False) as s:
                 cursor = self._command(
                     sock_info, cmd, slave_okay, session=s)["cursor"]
                 return CommandCursor(coll, cursor, sock_info.address, session=s,
-                                     explicit_session=session is not None)
+                                     explicit_session=session is not None,
+                                     batch_size=batch_size)
         else:
             coll = self["system.namespaces"]
+            if "name" in filter:
+                if not isinstance(filter["name"], string_type):
+                    raise TypeError("filter['name'] must be a string on MongoDB 2.6")
+                filter["name"] = coll.database.name + "." + filter["name"]
             res = _first_batch(sock_info, coll.database.name, coll.name,
-                               criteria, 0, slave_okay,
+                               filter, 0, slave_okay,
                                CodecOptions(), ReadPreference.PRIMARY, cmd,
-                               self.client._event_listeners, session=None)
+                               self.client._event_listeners, session=None, batch_size=batch_size)
             data = res["data"]
             cursor = {
                 "id": res["cursor_id"],
                 "firstBatch": data,
                 "ns": coll.full_name,
             }
-            return CommandCursor(coll, cursor, sock_info.address)
+            return CommandCursor(coll, cursor, sock_info.address,
+                                 batch_size=batch_size)
+    def list_collections(self, filter=None, session=None, batch_size=0):
+        """Get info about the collections in this database."""
+        with self.__client._socket_for_reads(
+                ReadPreference.PRIMARY) as (sock_info, slave_okay):
 
-    def collection_names(self, include_system_collections=True, session=None):
+             wire_version = sock_info.max_wire_version
+             results = self._list_collections(sock_info, slave_okay, filter=filter,
+                                           batch_size=batch_size, session=session)
+        # Iterating the cursor to completion may require a socket for getmore.
+        # Ensure we do that outside the "with" block so we don't require more
+        # than one socket at a time.
+        names = [result["name"] for result in results]
+        if wire_version <= 2:
+            # MongoDB 2.6 and older return index namespaces and collection
+            # namespaces prefixed with the database name.
+            names = [n[len(self.__name) + 1:] for n in names
+                     if n.startswith(self.__name + ".") and "$" not in n]
+        return names
```

Comment:
This needs to return the full `listCollections` response documents not just the names. Because we need to modify those documents before giving them back to the user, I think it makes sense for this method to be a generator:

```python
# run self._list_collections
for result in results:
    if wire_verison <= 2:
        name = result["name"]
        if "$" in name:
            continue
        result["name"] = name.split(".", 1)[1]
    yield result
```